src/lib/gl_engine/tvgGlTessellator.cpp

```diff
@@ -0,0 +1,2150 @@
+     *
+     *              v
+     *             / \
+     *          head  tail
+     *          /      \
+     *            ...
+     *
+     */
+    LinkedList<Edge> edge_below = {};
+
+    // left enclosing edge during sweep line
+    Edge *left = nullptr;
+    // right enclosing edge during sweep line
+    Edge *right = nullptr;
+
+    GlPoint point = {};
+
+    Vertex() = default;
+
+    Vertex(const GlPoint &p) : point(std::ceilf(p.x * 100.f) / 100.f, std::ceilf(p.y * 100.f) / 100.f)
+    {
+    }
+
+    ~Vertex() override = default;
+
+    bool isConnected() const
+    {
+        return edge_above.head || edge_below.head;
+    }
+
+    void insertAbove(Edge *e);
+    void insertBelow(Edge *e);
+};
+
+// we sort point by top first then left
+struct VertexCompare
+{
+    inline bool operator()(Vertex *v1, Vertex *v2)
+    {
+        return Compare(v1->point, v2->point);
+    }
+
+    static bool Compare(const GlPoint &a, const GlPoint &b);
+};
+
+// double linked list for all vertex in shape
+struct VertexList : public LinkedList<detail::Vertex>
+{
+    VertexList() = default;
+    VertexList(Vertex *head, Vertex *tail) : LinkedList(head, tail)
+    {
+    }
+
+    void insert(Vertex *v, Vertex *prev, Vertex *next);
+    void remove(Vertex *v);
+
+    void append(VertexList const &other);
+    void append(Vertex *v);
+    void prepend(Vertex *v);
+
+    void close();
+};
+
+struct Edge : public Object
+{
+    Vertex *top = nullptr;
+    Vertex *bottom = nullptr;
+
+    Edge *above_prev = nullptr;
+    Edge *above_next = nullptr;
+    Edge *below_prev = nullptr;
+    Edge *below_next = nullptr;
+
+    // left edge in active list during sweep line
+    Edge *left = nullptr;
+    // right edge in active list during sweep line
+    Edge *right = nullptr;
+
+    // edge list in polygon
+    Edge *right_poly_prev = nullptr;
+    Edge *right_poly_next = nullptr;
+    Edge *left_poly_prev = nullptr;
+    Edge *left_poly_next = nullptr;
+
+    // left polygon during sweep line
+    Polygon *left_poly = nullptr;
+    // right polygon during sweep line
+    Polygon *right_poly = nullptr;
+
+    bool used_in_left = false;
+    bool used_in_right = false;
+
+    int32_t winding = 1;
+
+    Edge(Vertex *top, Vertex *bottom, int32_t winding);
+
+    ~Edge() override = default;
+
+    // https://stackoverflow.com/questions/1560492/how-to-tell-whether-a-point-is-to-the-right-or-left-side-of-a-line
+    // return > 0 means point in left
+    // return < 0 means point in right
+    double sideDist(const GlPoint &p);
+
+    bool isRightOf(const GlPoint &p)
+    {
+        return sideDist(p) < 0.0;
+    }
+
+    bool isLeftOf(const GlPoint &p)
+    {
+        return sideDist(p) > 0.0;
+    }
+
+    // https://en.wikipedia.org/wiki/Line%E2%80%93line_intersection
+    bool intersect(Edge *other, GlPoint *point);
+
+    void recompute();
+
+    void setBottom(Vertex *v);
+
+    void setTop(Vertex *v);
+
+    void disconnect();
+
+private:
+    double le_a;
```

Comment:
Just to be certain, thorvg now utilizes a 32-bit coordinate system. Is this double precision mandatory?

Reply:
Yes, this is for precision.

Reply:
ok, makes sense.